roscala/src/main/scala/coop/rchain/rosette/prim/queue.scala

```diff
@@ -0,0 +1,204 @@
+package coop.rchain.rosette.prim
+
+import coop.rchain.rosette.Ob.ABSENT
+import coop.rchain.rosette.{Ctxt, Fixnum, Ob, Queue, RblBool, Tuple}
+import coop.rchain.rosette.macros.{checkArgumentMismatch, checkTypeMismatch}
+import coop.rchain.rosette.prim.Prim._
+
+object queue {
+
+  object queueNew extends Prim {
+    override val name: String = "queue-new"
+    override val minArgs: Int = 0
+    override val maxArgs: Int = 0
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Queue] =
+      Right(Queue.create())
+  }
+
+  object queueDepth extends Prim {
+    override val name: String = "queue-depth"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Fixnum] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      Right(q.nElems.asInstanceOf[Fixnum])
+    }
+  }
+
+  object queueIsEmpty extends Prim {
+    override val name: String = "queue-empty?"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, RblBool] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      Right(RblBool(q.isEmpty()))
+    }
+  }
+
+  object queueEnqueue extends Prim {
+    override val name: String = "queue-enqueue"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Queue] =
+      ctxt.argvec.elem.head match {
+        case q: Queue =>
+          val ob = ctxt.argvec.elem(1)
+          Right(q.enqueue(ob))
+        case _ =>
+          Left(ArgumentMismatch("The first element should be a queue"))
+      }
+  }
+
+  object queueDeQueue extends Prim {
+    override val name: String = "queue-dequeue"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] =
+      ctxt.argvec.elem.head match {
+        case q: Queue =>
+          if (q.isEmpty()) {
+            Left(QueueEmptyError)
+          } else {
+            Right(q.dequeue()._2)
+          }
+        case _ =>
+          Left(ArgumentMismatch("The first element should be a queue"))
+      }
+  }
+
+  object queueRead extends Prim {
+    override val name: String = "queue-read"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      if (q.isEmpty()) {
```

Comment:
Why not return `QueueEmptyError` here? Also you will return `Right(q.head)` even if `q.isEmpty()`.

Reply:
No, if q.isEmpty(), it will return a PrimError result, whether it's Absent or QueueEmptyError.

Reply:
You are missing an `else`. Right now you always return `Right(q.head)`.